### ISSUE TITLE
EF Core - Example of intercepting generate sql to support old SQL Server Versions

### ISSUE THREAD

**Issue by CorsairRO** [dotnet/efcore#30778]
## Ask a question

Hello,

I have many legacy apps that i would like to move from EF edmx to EF Core.
Those apps have lots of stored procedures with string concatenating columns.
Initially we have used the only available solution and that one was to use STUFF.
Then we made a CLR Aggregate dbo.StringConcatenateAggregate( string Value, string separator )
This one was way more performant to use then the STUFF.

Now if we want to get rid of stored procedures ( i have apps with 1000+ of them ) and move towards LINQ queries,
i am stuck there. 
I Know that string.Join translates automatically to STRING_AGG on newer SQL Server versions.
The problem is that we have to support Sql Server 2008, while STRING_AGG was introduced in 2016 i think.
As i understood from a previous question that some of you replied ( Arthur or Shay ), this is not supported and it is not in the plan to support by using EF Functions

In one of your Data Standups i think Arthur mentioned that it is possible to intercept expression trees to customize the generate SQL commands.
I would kindly ask if you have some example of how to do that. Or if you plan to have a Standup about it. Or just indicate me a starting point to look there on my own.

Basically i want if i find in Expression Tree that it wants to generate STRINGAGG to generate my dbo.StringConcatenateAggregate.
I dont care if it will throw me an exception if that aggregate is not in the database, it's my responsibility to have it in all my DB.
As my CLR has Exactly the same signature as STRINGAGG i just want to replace that and not touching any operands/parameters and so on. So in my opinion it should be very simple for me to achieve that.
But for that please give me a minimal starting point to look at.

Thank you very much in advance

### Include provider and version information

EF Core version: 7.0.5
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 7.0
Operating system: Windows 11
IDE: Visual Studio 2022 17.5

**Comment by roji** [comment]
For information on doing query interception, see the [EF 7.0 what's new docs](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#new-and-improved-interceptors-and-events).

However, I'm not quite sure what your planning... If you have some sort of custom version of STRING_AGG in your old databases and want to call that, I'd suggest simply making it named STRING_AGG, so that the normal EF-generated SQL will just work as-is. Otherwise, note that the query interceptor works at the very start of the query pipeline, on the raw, incoming LINQ expression tree; EF Core translates to STRING_AGG only much later.

In other words, if you need help here we're going to need to know more on exactly how you're planning to work around this.

**Comment by CorsairRO** [comment]
Hello,

Thank you for the answer.

The idea is that I cannot rename the CLR aggregate to STRING_AGG as I use SqlComparison to update a client DB at every update.
If the customer has Sql<2016, then it will work.
However if he has 2016+ the update script will complain that the STRING_AGG already exists implemented in SQL Server.

Or maybe I will investigate more to see if I can work like this.

Basically I am after writing a query like:

ctx.Order
                .Include( o=> o.OrderItem )
                .GroupBy( o=> o.OrderId )
                .Select new {
                                o.OrderId,
                                string.Join( oi.ItemNr )
                }
                .ToList();

And generate the SQL:
Select o.OrderId, dbo.StringConcatenate( o.ItemNr, “,”) AS Items
FROM Order as o
JOIN OrderItem as oi ON oi.OrderId = o.OrderId
GROUP BY o.OrderId

IF client has SQL>2016 this will work fine and will generate STRING_AGG.
Unfortunately I have lots of customers who are not willing to pay/upgrade SQL Server Versions, and we have to support 2008R2 as the minimum.

Like I said before in a previous question, the functionality exists in EF Core ALMOST how I want it to be.
In the way that i can use a Custom Function from EF Extensions in my linq queries.
As I looked there, there are some aggregates already defined.
The only problem is that they don’t support mapping a SqlCLR function aggregate.

I was assuming that I can intercept somehow the SQL Command String and just call a:
String.Replace( “STRING_AGG”, “dbo.StringConcatenate” ) before it is sent to SQL.

If not, I will investigate more how I can deal with renaming that object.

Thank you / Takk / Multumesc

Cosmin Palade
Senior Developer
Phone: +40 744 535 995
Email: ***@***.******@***.***>
Skype: cosmin.palade
Teams: ***@***.******@***.***>
Web: http://www.lundin.ro<http://www.lundin.ro/>

**Comment by roji** [comment]
> I was assuming that I can intercept somehow the SQL Command String and just call a:
> String.Replace( “STRING_AGG”, “dbo.StringConcatenate” ) before it is sent to SQL.

If this is what you're looking to do, a [command interceptor](https://learn.microsoft.com/en-us/ef/core/logging-events-diagnostics/interceptors#database-interception) can allow doing that. However, be aware that doing search-and-replace on SQL text is always brittle and can be error-prone (e.g. if STRING_AGG appears in some string somewhere).

Otherwise, the more "proper" way would be to override the SQL Server provider translation for string.Join to call your thing instead of STRING_AGG, but that's likely to be more complicated.

**Comment by CorsairRO** [comment]
Hello,

Ok, thank you for the answer.

I have just tried to create my aggregate as you suggested, naming it STRING_AGG
Unfortunately it doesn’t work even in SSMS 2008 to call it, as it has to be called: dbo.STRING_AGG
Calling a query in SSMS with dbo works fine, without it complains STRING_AGG doesn’t exist.

CREATE AGGREGATE [STRING_AGG]
***@***.*** [nvarchar](max), @separator [nvarchar](8))
RETURNS[nvarchar](max)
EXTERNAL NAME [MyCLRDll].[MyCLRDllNamespace.StringConcatenateAggregate]
GO

So  thank you for the tip with Command Interceptor, I think this is my way to go.

Please close the ticket otherwise, thank you for your answers, and have a nice day

Thank you / Takk / Multumesc

Cosmin Palade
Senior Developer
Phone: +40 744 535 995
Email: ***@***.******@***.***>
Skype: cosmin.palade
Teams: ***@***.******@***.***>
Web: http://www.lundin.ro<http://www.lundin.ro/>
